README.rst
=================
Batch Service API
=================

HTTP API for controlling, manipulating, and monitoring job batches.


docs/history.rst
Release History
===============

`Next Release`_
---------------

.. _Next Release: https://github.com/dave-shawley/finna-be-wight/compare/0.0.0...master


tests.py
from __future__ import print_function
import atexit
import logging
import urllib
import uuid

from pika import exceptions
from tornado import testing
import requests

from magician import rabbit


logging.getLogger('pika').setLevel(logging.DEBUG)
logging.getLogger('requests').setLevel(logging.DEBUG)


def amqp_control(method, *path, **kwargs):
    target = 'http://localhost:15672/api/{0}'.format(
        '/'.join(urllib.quote(segment, safe='') for segment in path))
    headers = kwargs.get('headers', {})
    headers['Content-Type'] = 'application/json'
    kwargs['headers'] = headers
    if 'auth' not in kwargs:
        kwargs['auth'] = ('guest', 'guest')
    response = requests.request(method, target, **kwargs)
    response.raise_for_status()
    return response


class RabbitConnectionConnectTests(testing.AsyncTestCase):

    @classmethod
    def setUpClass(cls):
        cls.virtual_host = '/' + uuid.uuid4().hex
        amqp_control('PUT', 'vhosts', cls.virtual_host)
        atexit.register(amqp_control, 'DELETE', 'vhosts', cls.virtual_host)

        amqp_control('PUT', 'permissions', cls.virtual_host, 'guest',
                     data='{"configure":".*","write":".*","read":".*"}')

    def setUp(self):
        super(RabbitConnectionConnectTests, self).setUp()
        amqp_url = 'amqp://guest:guest@localhost:5672/{0}'.format(
            urllib.quote(self.virtual_host, safe=''))
        self.rabbit_connection = rabbit.RabbitConnection(
            amqp_url, custom_ioloop=self.io_loop)

    @testing.gen_test
    def test_connection_blocks_until_connected(self):
        channel = yield self.rabbit_connection.connect()
        self.assertTrue(channel.is_open)

    @testing.gen_test
    def test_returned_channel_is_reused(self):
        channel = yield self.rabbit_connection.connect()
        first_channel_number = channel.channel_number
        self.rabbit_connection.return_channel(channel)

        channel = yield self.rabbit_connection.connect()
        self.assertEqual(channel.channel_number, first_channel_number)

    @testing.gen_test
    def test_closed_channel_is_not_reused(self):
        channel = yield self.rabbit_connection.connect()
        first_channel = channel
        self.rabbit_connection.return_channel(channel)
        first_channel.close()

        channel = yield self.rabbit_connection.connect()
        self.assertTrue(channel is not first_channel)

    @testing.gen_test
    def test_connection_failures_are_reported(self):
        new_connection = rabbit.RabbitConnection(
            'amqp://guest:guest@example.com:5672/%2F{0}?{1}'.format(
                uuid.uuid4().hex,
                urllib.urlencode({
                    'connection_attempts': 1,
                    'retry_delay': 0.001,
                    'socket_timeout': 0.001,
                }),
            ),
            custom_ioloop=self.io_loop,
        )
        try:
            yield new_connection.connect()
        except exceptions.AMQPError:
            return
        self.fail('Expected exception to be raised')

    @testing.gen_test
    def test_connecting_on_closed_connection(self):
        def print_message(msg):
            def printer(*args):
                logging.debug(msg + ': %s', args)
            return printer

        first_channel = yield self.rabbit_connection.connect()
        self.rabbit_connection._connection.add_on_close_callback(
            print_message('CONNECTION CLOSED'))
        first_channel.add_on_close_callback(
            print_message('FIRST CHANNEL CLOSED'))
        logging.debug('CLOSING CONNECTION')

        self.io_loop.run_sync(self.rabbit_connection._connection.close)

        logging.debug('ATTEMPTING SECOND OPEN')
        second_channel = yield self.rabbit_connection.connect()
        self.assertEqual(first_channel, second_channel)


tests/__init__.py
import os


def setup_package():
    with open('build/test-environment', 'r') as env_file:
        for line in env_file:
            if '#' in line:
                line = line[:line.index('#')]
            name, sep, value = line.partition('=')
            name, value = name.strip(), value.strip()
            if name and sep:
                if value.startswith('"', "'") and value[0] == value[-1]:
                    value = value[1:-1].strip()
                if value:
                    os.environ[name] = value
                else:
                    os.environ.pop(name, None)


magician/rabbit.py
from __future__ import print_function

import json
import logging
import time

from tornado import concurrent, gen
import pika
import pika.adapters
import pika.exceptions
import pika.spec


class RabbitConnection(object):

    def __init__(self, amqp_url, custom_ioloop=None):
        super(RabbitConnection, self).__init__()
        self.amqp_url = amqp_url
        self._connection = None
        self._future = None
        self._channels = []
        self._custom_io_loop = custom_ioloop
        self._logger = logging.getLogger('RabbitConnection')
        self._logger.debug('connection created for %s', self.amqp_url)

    @gen.coroutine
    def connect(self):
        """
        Asynchronously connect to rabbit.

        :return: a :class:`pika.Channel` instance
        :raises pika.exceptions.AMQPError:
            if a protocol error occurs when connecting to the
            AMQP broker

        """
        self._future = concurrent.Future()
        if self._connection is None:
            self._logger.debug('connecting to %s', self.amqp_url)

            def on_connection_failure(*args):
                self._logger.error('connection failure - %s', args)
                self._future.set_exception(pika.exceptions.AMQPConnectionError)

            pika.adapters.TornadoConnection(
                pika.URLParameters(self.amqp_url),
                on_open_callback=self.open_channel,
                on_open_error_callback=on_connection_failure,
                custom_ioloop=self._custom_io_loop,
            )
        else:
            self._logger.debug('already connected, scanning channel cache')
            while self._channels:
                channel = self._channels.pop()
                if channel.is_open:
                    self._logger.debug('returning cached channel %s (%d)',
                                       channel, channel.channel_number)
                    raise gen.Return(channel)

            self._logger.debug('generating channel')
            self._connection.channel(
                on_open_callback=self.handle_channel_opened)

        self._logger.debug('waiting on connection')
        channel = yield gen.YieldFuture(self._future)
        self._logger.debug('received channel %s (%d)',
                           channel, channel.channel_number)
        raise gen.Return(channel)

    def open_channel(self, connection):
        self._logger.debug('connected to %s, asking for channel', connection)
        self._connection = connection
        connection.add_on_close_callback(self.handle_connection_closed)
        connection.channel(on_open_callback=self.handle_channel_opened)

    def handle_connection_closed(self, *args):
        self._logger.info('connection closed - %s', args)
        self._connection = None
        self._channels = []

    def handle_channel_opened(self, channel):
        self._logger.debug('channel opened: %s (%d)',
                           channel, channel.channel_number)
        channel.add_on_close_callback(self.handle_channel_closed)
        self._future.set_result(channel)
        self._future = None

    def handle_channel_closed(self, dead_channel, *args):
        self._logger.debug('channel %s (%d) closed - %s',
                           dead_channel, dead_channel.channel_number, args)
        for index, channel in self._channels:
            if channel.channel_number == dead_channel.channel_number:
                self._logger.debug('removing channel %d from internal list',
                                   channel.channel_number)
                del self._channels[index]
                break

    def return_channel(self, channel):
        if channel is not None:
            if channel.is_open:
                self._logger.debug('channel %d returned',
                                   channel.channel_number)
                self._channels.append(channel)
            else:
                self._logger.debug('closed channel %d returned, discarding',
                                   channel.channel_number)


class PublisherMixin(object):

    def __init__(self, *args, **kwargs):
        super(PublisherMixin, self).__init__(*args, **kwargs)
        self.channel = None
        self.connection = None

    @gen.coroutine
    def prepare(self):
        super(PublisherMixin, self).prepare()
        self.rabbit_url = 'amqp://guest:guest@localhost:5672/%2F'
        rabbit_holes = getattr(self.application, '_rabbit_holes', None)
        if rabbit_holes is None:
            rabbit_holes = self.application._rabbit_holes = {}

        try:
            self.connection = rabbit_holes[self.rabbit_url]
        except KeyError:
            self.connection = RabbitConnection(self.rabbit_url)
            rabbit_holes[self.rabbit_url] = self.connection

        self.channel = yield self.connection.connect()

    def on_finish(self):
        if self.channel is not None:
            self.connection.return_channel(self.channel)

    @gen.coroutine
    def publish(self, exchange, routing_key, body, **properties):

        amqp_properties = pika.BasicProperties(
            content_type=properties.get('content_type', None),
            content_encoding=properties.get('content_encoding', None),
            headers=properties.get('headers', None),
            delivery_mode=properties.get('delivery_mode', None),
            priority=properties.get('priority', None),
            correlation_id=properties.get('correlation_id', None),
            reply_to=properties.get('reply_to', None),
            expiration=properties.get('expiration', None),
            message_id=properties.get('message_id', None),
            timestamp=properties.get('timestamp', time.time()),
            type=properties.get('type', None),
            user_id=properties.get('user_id', None),
            app_id=properties.get('app_id', None),
            cluster_id=properties.get('cluster_id', None),
        )

        if amqp_properties.content_type is not None:
            content_type = amqp_properties.content_type
            if ';' in content_type:
                content_type = content_type[:content_type.index(';')].strip()
            if content_type.endswith('json'):
                body = json.dumps(body).encode('utf-8')

        future = concurrent.Future()

        def channel_closed(*args):
            print('CLOSED', args)
            future.set_result(500)

        def delivery_callback(*args):
            print('CONFIRMED', args)
            future.set_result(200)

        # self.channel.add_on_close_callback(channel_closed)
        self.channel.callbacks.add(
            prefix=self.channel.channel_number,
            key='_on_channel_close',
            callback=channel_closed,
            one_shot=True,
            only_caller=self.channel,
        )

        # self.channel.confirm_delivery(delivery_callback)
        self.channel.callbacks.add(
            self.channel.channel_number,
            pika.spec.Basic.Ack,
            delivery_callback,
            one_shot=True,
        )
        self.channel.callbacks.add(
            self.channel.channel_number,
            pika.spec.Basic.Nack,
            delivery_callback,
            one_shot=True,
        )
        self.channel.confirm_delivery()

        print('PUBLISHING F`REAL')
        self.channel.basic_publish(
            exchange, routing_key, body,
            mandatory=properties.get('mandatory', False),
            properties=amqp_properties,
        )

        status_code = yield gen.YieldFuture(future)
        raise gen.Return(status_code)


magician/parameters.py
import json

from tornado import web


_DEFAULT_ARGUMENT = object()


class MissingParameter(web.HTTPError):

    def __init__(self, name):
        super(MissingParameter, self).__init__(
            status_code=400,
            log_message='Missing parameter {0}'.format(name),
        )
        self.parameter_name = name


class InvalidParameter(web.HTTPError):

    def __init__(self, name, value):
        super(InvalidParameter, self).__init__(
            status_code=400,
            log_message='"{0}" is not a valid value for {1}'.format(
                str(value), name),
        )
        self.parameter_name = name
        self.parameter_value = value


def extract_parameter(name, params, default, factory):
    try:
        value = params[name]
        if factory is not None:
            value = factory(value)
    except ValueError:
        raise InvalidParameter(name, value)
    except KeyError:
        if default is _DEFAULT_ARGUMENT:
            raise MissingParameter(name)
        value = default
    return value


class ParameterMixin(object):

    def prepare(self):
        super(ParameterMixin, self).prepare()
        self._json = None

    def path_parameter(self, name, default=_DEFAULT_ARGUMENT, factory=None):
        return extract_parameter(name, self.path_kwargs, default, factory)

    def json_parameter(self, name, default=_DEFAULT_ARGUMENT, factory=None):
        if self._json is None:
            self._json = json.loads(self.request.body)
        return extract_parameter(name, self._json, default, factory)


magician/web.py
from __future__ import print_function
import json
try:
    from urllib import parse
except ImportError:
    import urllib as parse

from pika import exceptions
from tornado import gen, httpclient, ioloop, web

from magician import parameters, rabbit


def to_bool(value):
    if value.lower() == 'false':
        return False
    if value.lower() == 'true':
        return True
    try:
        return bool(int(value))
    except ValueError:
        pass
    raise ValueError(value)


class RabbitCreator(parameters.ParameterMixin, web.RequestHandler):

    def prepare(self):
        super(RabbitCreator, self).prepare()
        self.client = httpclient.AsyncHTTPClient(defaults={
            'auth_username': 'guest', 'auth_password': 'guest',
        })

    def _call_api(self, method, *path, **kwargs):
        headers = kwargs.pop('headers', {})
        headers['Content-Type'] = 'application/json'

        body = kwargs.get('body', None)
        if body is not None:
            kwargs['body'] = json.dumps(body)

        request = httpclient.HTTPRequest(
            self._rabbit_url(*path), method=method,
            headers=headers, **kwargs)
        print('REQUEST', request, request.url, request.headers)
        return self.client.fetch(request)

    @staticmethod
    def _rabbit_url(*path):
        url = 'http://localhost:15672/api/{0}'.format('/'.join(
            parse.quote(segment, safe='') for segment in path))
        return url

    @gen.coroutine
    def post(self, **_):
        response = None
        item_type = self.path_parameter('itemtype')
        if item_type in ('binding', 'queue'):
            response = yield self._call_api(
                'PUT',
                'queues',
                self.json_parameter('virtual_host', default='/'),
                self.json_parameter('queue_name'),
                body={
                    'auto_delete': False,
                    'durable': self.json_parameter('durable',
                                                   default=False,
                                                   factory=to_bool),
                },
            )
            if response.code >= 400:
                raise web.HTTPError(response.code)

        if item_type in ('binding', 'exchange'):
            response = yield self._call_api(
                'PUT',
                'exchanges',
                self.json_parameter('virtual_host', default='/'),
                self.json_parameter('exchange_name'),
                body={
                    'type': self.json_parameter('exchange_type',
                                                default='topic'),
                    'durable': self.json_parameter('durable',
                                                   default=False,
                                                   factory=to_bool),
                },
            )
            if response.code >= 400:
                raise web.HTTPError(response.code)

        if item_type == 'binding':
            response = yield self._call_api(
                'POST',
                'bindings',
                self.json_parameter('virtual_host', default='/'),
                'e', self.json_parameter('exchange_name'),
                'q', self.json_parameter('queue_name'),
                body={
                    'routing_key': self.json_parameter('routing_key'),
                },
            )
            if response.code >= 400:
                raise web.HTTPError(response.code)

        if response is None:
            raise web.HTTPError(500)

        self.set_status(200)


class RabbitPublisher(
        rabbit.PublisherMixin, parameters.ParameterMixin, web.RequestHandler):

    @gen.coroutine
    def get(self, *args, **kwargs):
        try:
            print('PUBLISHING')
            result = yield self.publish('po', 'foo.mail', '{"name":"blah"}')
            print('RETURNING IT ALL')
            self.set_status(result)
        except exceptions.AMQPError:
            self.set_status(500)


application = web.Application(
    [
        (r'/(?P<itemtype>queue)', RabbitCreator),
        (r'/(?P<itemtype>exchange)', RabbitCreator),
        (r'/(?P<itemtype>binding)', RabbitCreator),
        (r'/publish', RabbitPublisher),
    ],
    debug=True,
)


if __name__ == '__main__':
    application.listen(8000)
    ioloop.IOLoop.instance().start()
